File: BST/simple_bst/lingkerio/C-version/correct.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include "bst.h"  // Assuming you put the BST implementation in bst.h

#define MAX_LINE_LENGTH 100

// Structure to hold an operation and its value
typedef struct {
    char operation[MAX_LINE_LENGTH];
    int value;
} Operation;

// Function to read operations from a CSV file into an array
int load_operations(const char* filename, Operation** operations) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        perror("Error opening file");
        return -1;
    }

    int capacity = 100;  // Initial capacity for the operations array
    int count = 0;
    *operations = (Operation*)malloc(capacity * sizeof(Operation));

    char line[MAX_LINE_LENGTH];
    while (fgets(line, sizeof(line), file)) {
        if (count >= capacity) {
            capacity *= 2;
            *operations = (Operation*)realloc(*operations, capacity * sizeof(Operation));
        }
        if (sscanf(line, "%[^,],%d", (*operations)[count].operation, &(*operations)[count].value) == 2) {
            count++;
        } else {
            fprintf(stderr, "Warning: Malformed line: %s", line);
        }
    }

    fclose(file);
    return count;
}

// Function to process a single operation and record the result
BSTNode* process_operation(BSTNode* root, const char* operation, int value, FILE* output_file) {
    if (strcmp(operation, "insert") == 0) {
        root = insert(root, value);
        fprintf(output_file, "insert,%d\n", value);
    } else if (strcmp(operation, "search") == 0) {
        bool found = search(root, value);
        fprintf(output_file, "search,%d,%s\n", value, found ? "true" : "false");
    } else if (strcmp(operation, "delete") == 0) {
        BSTNode* node_to_delete = search(root, value) ? root : NULL;
        root = delete(root, value);
        if (node_to_delete) {
            fprintf(output_file, "delete,%d\n", value);
        } else {
            fprintf(output_file, "delete,%d,null\n", value);
        }
    } else if (strcmp(operation, "findmin") == 0) {
        BSTNode* min_node = find_min(root);
        if (min_node) {
            fprintf(output_file, "findmin,%d\n", min_node->value);
        } else {
            fprintf(output_file, "findmin,null\n");
        }
    } else if (strcmp(operation, "findmax") == 0) {
        BSTNode* max_node = find_max(root);
        if (max_node) {
            fprintf(output_file, "findmax,%d\n", max_node->value);
        } else {
            fprintf(output_file, "findmax,null\n");
        }
    } else {
        fprintf(stderr, "Warning: Unknown operation: %s\n", operation);
    }
    return root;
}

// Function to execute the BST operations and record the results
void execute_operations(Operation* operations, int count, const char* output_filename) {
    FILE* output_file = fopen(output_filename, "w");
    if (output_file == NULL) {
        perror("Error opening output file");
        return;
    }

    BSTNode* root = NULL;
    for (int i = 0; i < count; i++) {
        root = process_operation(root, operations[i].operation, operations[i].value, output_file);
    }

    free_bst(root);
    fclose(output_file);
}

int main() {
    const char* input_filename = "../../../benchmark/benchmark_operations.csv";
    const char* output_filename = "correctness.csv";

    Operation* operations;
    int count = load_operations(input_filename, &operations);
    if (count < 0) {
        return 1;
    }

    execute_operations(operations, count, output_filename);

    free(operations);
    return 0;
}

File: BST/simple_bst/lingkerio/Python-version/correct.py
import sys
import Bst

sys.setrecursionlimit(3000)

def load_operations(filename):
    operations = []
    with open(filename, "r") as file:
        for line in file:
            operation, value_str = line.strip().split(',')
            value = int(value_str)
            operations.append((operation, value))
    return operations

def process_operation(bst, operation, value, output_file):
    if operation == "insert":
        bst.insert(value)
        output_file.write(f"insert,{value}\n")
    elif operation == "search":
        found = bst.search(value)
        output_file.write(f"search,{value},{'true' if found else 'false'}\n")
    elif operation == "delete":
        node_to_delete = bst.search(value)
        bst.delete(value)
        if node_to_delete:
            output_file.write(f"delete,{value}\n")
        else:
            output_file.write(f"delete,{value},null\n")
    elif operation == "findmin":
        min_value = bst.find_min()
        if min_value is not None:
            output_file.write(f"findmin,{min_value}\n")
        else:
            output_file.write(f"findmin,null\n")
    elif operation == "findmax":
        max_value = bst.find_max()
        if max_value is not None:
            output_file.write(f"findmax,{max_value}\n")
        else:
            output_file.write(f"findmax,null\n")
    else:
        output_file.write(f"Invalid operation: {operation}\n")

if __name__ == "__main__":
    filename = "../../../benchmark/benchmark_operations.csv"
    output_filename = "correctness.csv"
    
    operations = load_operations(filename)
    
    bst = Bst.BST()
    with open(output_filename, "w") as output_file:
        for operation, value in operations:
            process_operation(bst, operation, value, output_file)
    
    print("Correctness test completed. Results are saved in correctness.csv.")